tinyyaml.lua

```diff
@@ -259,561 +189,641 @@ local function checkdupekey(map, key)
+        end
         line = ssub(line, 2)
+      elseif c == '}' then
+        if stack[#stack].t == '{' then
+          if #stack == 1 then break end
+          stack[#stack].t = '}'
+          line = ssub(line, 2)
+        else
+          line = ','..line
+        end
+      elseif c == ']' then
+        if stack[#stack].t == '[' then
+          if #stack == 1 then break end
+          stack[#stack].t = ']'
+          line = ssub(line, 2)
+        else
+          line = ','..line
+        end
       else
-        line = ','..line
-      end
-    else
-      local s, rest = parsestring(line, ',{}[]')
-      if not s then
-        error('invalid flowstyle line: '..line)
+        local s, rest = parsestring(line, ',{}[]')
+        if not s then
+          error('invalid flowstyle line: '..line)
+        end
+        tinsert(stack, {v=s, t='s'})
+        line = rest
       end
-      tinsert(stack, {v=s, t='s'})
-      line = rest
     end
+    return stack[1].v, line
   end
-  return stack[1].v, line
-end
 
-local function parseblockstylestring(line, lines, indent)
-  if #lines == 0 then
-    error("failed to find multi-line scalar content")
-  end
-  local s = {}
-  local firstindent = -1
-  local endline = -1
-  for i = 1, #lines do
-    local ln = lines[i]
-    local idt = countindent(ln)
-    if idt <= indent then
-      break
+  local function parseblockstylestring(line, lines, indent)
+    if #lines == 0 then
+      error("failed to find multi-line scalar content")
     end
-    if ln == '' then
-      tinsert(s, '')
-    else
-      if firstindent == -1 then
-        firstindent = idt
-      elseif idt < firstindent then
+    local s = {}
+    local firstindent = -1
+    local endline = -1
+    for i = 1, #lines do
+      local ln = lines[i]
+      local idt = countindent(ln)
+      if idt <= indent then
         break
       end
-      tinsert(s, ssub(ln, firstindent + 1))
+      if ln == '' then
+        tinsert(s, '')
+      else
+        if firstindent == -1 then
+          firstindent = idt
+        elseif idt < firstindent then
+          break
+        end
+        tinsert(s, ssub(ln, firstindent + 1))
+      end
+      endline = i
     end
-    endline = i
-  end
 
-  local striptrailing = true
-  local sep = '\n'
-  local newlineatend = true
-  if line == '|' then
-    striptrailing = true
-    sep = '\n'
-    newlineatend = true
-  elseif line == '|+' then
-    striptrailing = false
-    sep = '\n'
-    newlineatend = true
-  elseif line == '|-' then
-    striptrailing = true
-    sep = '\n'
-    newlineatend = false
-  elseif line == '>' then
-    striptrailing = true
-    sep = ' '
-    newlineatend = true
-  elseif line == '>+' then
-    striptrailing = false
-    sep = ' '
-    newlineatend = true
-  elseif line == '>-' then
-    striptrailing = true
-    sep = ' '
-    newlineatend = false
-  else
-    error('invalid blockstyle string:'..line)
-  end
+    local striptrailing = true
```

Comment:
Look like most of the change is indent? Please keep the original indent.

Reply:
Um… if i include a function into a closure this should increase the indent, shouldn't it? 

I could unindent it so the changes are easier visible and move every function in a single commit. 

Generally I did the following:

```
-- all code of tinyyaml.lua except parse functions and parse

local function parse(source, options)

  if options == nil then
    options = {}
  end

  local function parse……
    -- original code with increated indent
  end

  local function parse_inner(source)
    -- code of parse, depends on parsescalar, which depends on options
  end

  return parse_inner(source)

end

```

Reply:
What about using options as a module-level variable?

Reply:
that's what I did before (if you want to have a look that's here: https://github.com/TeXhackse/lua-tinyyaml/commit/2049ad66e135f81f60a7089ffad4e4b16bebbca4)

 but @Witiko mentioned this might lead to an issue if one is using multiple instances of the module.  https://github.com/Witiko/markdown/issues/114#issuecomment-1003151543

I am not a Lua pro but mostly hacking TeX/LaTeX. So I adopted his suggestion of a closure. In case the commit mentioned above would be better for you I can close this one and convert that one to a PR.

Reply:
> What about using options as a module-level variable?

@spacewander This is fine if the `tinyyaml` is only used in one place. However, as soon as several libraries in a single Lua process use tinyyaml, we will be in a world of grief. Consider the following example:

Library A imports `tinyyaml`, sets the module-level `options` variable, and then assumes that it won't change:

``` lua
yaml = require("tinyyaml")
yaml.options.timestamps = false
```

This seems to work as long as there is no other library that uses tinyyaml. However, now there is also library B that does not know about the `options` variable, but depends on timestamp parsing, which used to be the default behavior. After A changes `options.timestamps`, B breaks. Next, there is library C that does the right thing and changes `options.timestamps` before every call to `parse`. C breaks both A and B.

All of these can be easily solved by making `options` from a module-level variable to a function argument as implemented by @TeXhackse. I agree that it may make sense to remove the indent and squash the commits to see what is going on.

Reply:
ouch - you are right. I actually was close to this thought when I did the adjustments yesterday but didn't get to it.